aws-sdk-core/lib/aws-sdk-core/endpoint_provider.rb

```diff
@@ -2,30 +2,52 @@ module Aws
   # @api private
   class EndpointProvider
 
-    # @api private
     PATH = File.join(File.dirname(__FILE__), '..', '..', 'endpoints.json')
-
-    # @api private
-    RULES = Json.load_file(PATH)['endpoints']
+    RULES = Json.load_file(PATH)
 
     class << self
 
       def resolve(region, service)
-        keys(region, service).each do |key|
-          if match = RULES[key]
-            return expand(match['endpoint'], region.to_s, service.to_s)
-          end
-        end
+        "https://" + endpoint_for(region, service)
       end
 
       private
+      def endpoint_for(region, service)
+        partition = get_partition(region)
+        if partition.nil?
+          return "#{service}.#{region}.amazonaws.com"
+        end
+        endpoint = default_endpoint(partition, service, region)
+        if service_cfg = partition["services"][service]
+          if service_cfg["isRegionalized"] == false
+            if partition_endpoint = service_cfg["partitionEndpoint"]
+              region = partition_endpoint
+            end
+          end
+          if endpoint_cfg = service_cfg["endpoints"][region]
+            if host = endpoint_cfg["hostname"]
+              endpoint = host
+            end
+          end
+        end
+        endpoint
+      end
 
-      def keys(region, service)
-        ["#{region}/#{service}", "#{region}/*", "*/#{service}", "*/*"]
+      def default_endpoint(partition, service, region)
+        hostname_template = partition["defaults"]["hostname"]
+        hostname_template.sub(
+          '{region}', region
+        ).sub(
+          '{service}', service
+        ).sub(
+          '{dnsSuffix}', partition["dnsSuffix"]
+        )
       end
 
-      def expand(pattern, region, service)
-        'https://' + pattern.sub('{region}', region).sub('{service}', service)
+      def get_partition(region)
+        RULES["partitions"].find do |p|
+          region.match(p["regionRegex"])
+        end
```

Comment:
The #get_partition method could own the logic of returning 'aws' as the default partion if the region does not match the regex of any known partition. This removes the need for the nil check above.